src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,1664 @@
+					 image_extents_summary.last_block),
+			     new C_OnFinisher(write_req, &m_on_persist_finisher));
+  {
+    uint64_t buffer_offset = 0;
+    Mutex::Locker locker(m_lock);
+    /* TODO: If there isn't space for this whole write, defer it */
+    if (m_free_log_entries < write_req->image_extents.size()) {
+      ldout(cct, 6) << "wait for " << write_req->image_extents.size() << " log entries" << dendl;
+    } else {
+      for (auto &extent : write_req->image_extents) {
+	/* operation->on_write_persist connected to m_prior_log_entries_persisted Gather */
+	WriteLogOperation *operation =
+	  new WriteLogOperation(set, extent.first, extent.second);
+	set->operations.emplace_back(operation);
+	log_entries.emplace_back(operation->log_entry);
+
+	/* Reserve both log entry and data block space */
+	m_free_log_entries--;
+
+	/* Allocate data blocks */
+	operation->log_entry->ram_entry.has_data = 1;
+	/* TODO: Handle failure */
+	unsigned int alloc_bytes = MIN_WRITE_ALLOC_SIZE;
+	if (operation->log_entry->ram_entry.write_bytes > alloc_bytes) {
+	  alloc_bytes = operation->log_entry->ram_entry.write_bytes;
+	}
+	operation->log_entry->ram_entry.write_data = pmemobj_reserve(m_log_pool,
+								     &operation->buffer_alloc_action,
+								     alloc_bytes,
+								     0 /* Object type */);
+	assert(!TOID_IS_NULL(operation->log_entry->ram_entry.write_data));
+	operation->log_entry->pmem_buffer = D_RW(operation->log_entry->ram_entry.write_data);
+	operation->log_entry->ram_entry.sync_gen_number = m_current_sync_gen;
+	if (set->m_persist_on_flush) {
+	  /* Persist on flush. Sequence #0 is never used. */
+	  operation->log_entry->ram_entry.write_sequence_number = 0;
+	} else {
+	  /* Persist on write */
+	  operation->log_entry->ram_entry.write_sequence_number = ++m_last_op_sequence_num;
+	  operation->log_entry->ram_entry.sequenced = 1;
+	}
+	operation->log_entry->ram_entry.sync_point = 0;
+	operation->log_entry->ram_entry.unmap = 0;
+	operation->bl.substr_of(write_req->bl, buffer_offset,
+				operation->log_entry->ram_entry.write_bytes);
+	buffer_offset += operation->log_entry->ram_entry.write_bytes;
+	ldout(cct, 6) << "operation=[" << *operation << "]" << dendl;
+      }
+    }
+  }
+
+  m_async_op_tracker.start_op();
+  write_req->_on_finish =
+    new FunctionContext([this, write_req](int r) {
+	complete_write_req(write_req, r);
+	m_async_op_tracker.finish_op();
+      });
+
+  /* All extent ops subs created */
+  set->m_extent_ops->activate();
+  
+  /* Write data */
+  for (auto &operation : set->operations) {
+    bufferlist::iterator i(&operation->bl);
+    ldout(cct, 6) << operation->bl << dendl;
+    i.copy((unsigned)operation->log_entry->ram_entry.write_bytes, (char*)operation->log_entry->pmem_buffer);
+  }
+
+  m_blocks_to_log_entries.add_entries(log_entries);
+  
+  if (set->m_persist_on_flush) {
+    /* 
+     * We're done with the caller's buffer, and not guaranteeing
+     * persistence until the next flush. The block guard for this
```

Comment:
... so write-after-write to the same block(s) will potentially hang waiting for the guard to be released?

Reply:
No, the guard is release when write_req itself is finally completed. That occurs down the twisty maze of chained Context objects when the log entry for the write is persisted. (It may also be gated by the following sync point, but if it is that's a relic of the previous replication scheme.) The user request (the on_finish from ::aio_write) will be completed then unless it has already been completed here.

When we're using SSDs for the RWL, we can't get away with this early completion anyway. I need to make it selectable, and possibly force it off unless we know the local replica is really 3DXPoint. It's unclear how important this early completion is to performance, so I don't have a handle yet on whether it's worth the trouble.

The risk of this early complete path is that if the local RWL replica fails before persistence completes, the write is lost. The only remaining copy of that write data is in the address space of the local replica device, which will be unreadable when that device fails. That risk is acceptable if you assume that 3DXPoint won't fail any more often than DRAM, and when it does you'll get a machine check your OS probably can't survive. That means the local replica and the writing node fail together (at least for now). There is no incremental risk because we're already talking about persist-on-flush mode, where writes are at risk until the following flush completes.

We can all probably think of ways to avoid or handle that machine check So this wasn't always going to be true. For now I though it was reasonable simplifying assumption.

When the local replica is on an SSD, this doesn't make sense. The node will survive the failure of the SSD. An HA write back cache is expected to survive the failure of a cache replica, so failing the RBD volume when the local replica fails would be unacceptable. 

So this behavior needs to be disabled when the local replica isn't an NVDIMM, or we need a copy of the payload data in case the address space of the local replica becomes unreadable. We could copy small writes and defer completion of large writes to avoid that.